test/examples/fortran_example.f90

```diff
@@ -0,0 +1,324 @@
+    density(5, 3, ipoint) = 11.0d0  ! first order (3)
+    density(5, 4, ipoint) = 12.0d0  ! first order (3)
+    density(6, 1, ipoint) = 5.0d0   ! second order (depending on (1) and (3))
+    density(6, 2, ipoint) = 6.0d0   ! second order (depending on (1) and (3))
+    density(6, 3, ipoint) = 7.0d0   ! second order (depending on (1) and (3))
+    density(6, 4, ipoint) = 8.0d0   ! second order (depending on (1) and (3))
+    density(7, 1, ipoint) = 9.0d0   ! second order (depending on (2) and (3))
+    density(7, 2, ipoint) = 10.0d0  ! second order (depending on (2) and (3))
+    density(7, 3, ipoint) = 11.0d0  ! second order (depending on (2) and (3))
+    density(7, 4, ipoint) = 12.0d0  ! second order (depending on (2) and (3))
+    density(8, 1, ipoint) = 0.0d0   ! third order (depending on (1), (2) and (3))
+    density(8, 2, ipoint) = 0.0d0   ! third order (depending on (1), (2) and (3))
+    density(8, 3, ipoint) = 0.0d0   ! third order (depending on (1), (2) and (3))
+    density(8, 4, ipoint) = 0.0d0   ! third order (depending on (1), (2) and (3))
+  end do
+
+  res = derivative(id, &
+                   num_grid_points, &
+                   num_density_variables, &
+                   vector_length, &
+                   density)
+  deallocate(density)
+
+  ! compare with reference
+  call assert((abs(47.091223089835331d0 - res) < 1.0d-6), &
+              "derivatives do not match reference numbers")
+
+
+  !-----------------------------------------------------------------------------
+  ! we are done and can release the memory
+
+  call xc_free_functional(id)
+
+  print *, 'Kernel test has ended properly!'
+
+contains
+
+  real(8) function derivative(id, &
+                              num_grid_points, &
+                              num_density_variables, &
+                              vector_length, &
+                              density)
+    ! computes the derivative and takes care of offsetting
+
+    integer, intent(in) :: id
+    integer, intent(in) :: num_grid_points
+    integer, intent(in) :: num_density_variables
+    integer, intent(in) :: vector_length
+    real(8), intent(in) :: density(vector_length, num_density_variables, num_grid_points)
+
+    real(8), allocatable :: input_array(:, :)
+    real(8), allocatable :: output_array(:, :)
+
+    integer :: ipoint
+
+    allocate(input_array(num_density_variables*vector_length, num_grid_points))
+    allocate(output_array(vector_length, num_grid_points))
+
+    ! put the densities into the right places
+    ! along the input array
+    do ipoint = 1, num_grid_points
+      input_array(:, ipoint) = (/density(:, 1, ipoint), &
+                                 density(:, 2, ipoint), &
+                                 density(:, 3, ipoint), &
+                                 density(:, 4, ipoint)/)
+    end do
+
+    call xc_eval(id, num_grid_points, input_array, output_array)
+    derivative = output_array(vector_length, 1)
+
```

Comment:
Do I read this right that the output of xcfun is indeed ALL possible contractions, given the order and the corresponding densities, but then you only return one of them for the test (the last one)?

Reply:
The interface returns a Taylor series expansion and we pick here one element out of it. I will add a comment on this.